File: arey/platform/llm.py
"""Large language model abstraction."""
# pyright: strict, reportAny=false

from typing import Literal

from arey.core import CompletionModel, ModelConfig
from arey.platform._llama import LlamaBaseModel
from arey.platform._ollama import OllamaBaseModel
from arey.platform._openai import OpenAIBaseModel

ModelType = Literal["openai", "ollama", "gguf"]


def get_completion_llm(model_config: ModelConfig) -> CompletionModel:
    """Get a completion AI model."""
    if model_config.type == "ollama":
        return OllamaBaseModel(model_config)
    if model_config.type == "openai":
        return OpenAIBaseModel(model_config)

    return LlamaBaseModel(model_config)

File: src/platform/llm.rs
use crate::core::completion::CompletionModel;
use crate::core::model::ModelProvider;
use crate::platform::{llama, openai};
use anyhow::Result;

#[derive(thiserror::Error, Debug)]
pub enum ModelInitError {
    // #[error("Unsupported model type: {0}")]
    // UnsupportedType(String),
    // #[error("Initialization error: {0}")]
    // InitError(String),
}

pub fn get_completion_llm(
    model_config: crate::core::model::ModelConfig,
) -> Result<Box<dyn CompletionModel + Send + Sync>> {
    match model_config.provider {
        ModelProvider::Gguf => {
            let model = llama::LlamaBaseModel::new(model_config)?;
            Ok(Box::new(model))
        }
        ModelProvider::Openai => {
            let model = openai::OpenAIBaseModel::new(model_config)?;
            Ok(Box::new(model))
        }
    }
}
